pkg/kubelet/kuberuntime/kuberuntime_sandbox.go

```diff
@@ -0,0 +1,166 @@
+		hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+		if err != nil {
+			return nil, err
+		}
+		podSandboxConfig.Hostname = &hostname
+	}
+
+	cgroupParent := ""
+	podResourceLimits := make(map[api.ResourceName]int64)
+	podResourceRequests := make(map[api.ResourceName]int64)
+	portMappings := []*runtimeApi.PortMapping{}
+	for _, c := range pod.Spec.Containers {
+		for name, limit := range c.Resources.Limits {
+			if l, ok := podResourceLimits[name]; ok {
+				podResourceLimits[name] = l + limit.MilliValue()
+			} else {
+				podResourceLimits[name] = limit.MilliValue()
+			}
+		}
+		for name, req := range c.Resources.Requests {
+			if l, ok := podResourceRequests[name]; ok {
+				podResourceRequests[name] = l + req.MilliValue()
+			} else {
+				podResourceRequests[name] = req.MilliValue()
+			}
+		}
+
+		opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, &c, podIP)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, port := range opts.PortMappings {
+			hostPort := int32(port.HostPort)
+			containerPort := int32(port.ContainerPort)
+			protocol := toRuntimeProtocol(port.Protocol)
+			portMappings = append(portMappings, &runtimeApi.PortMapping{
+				HostIp:        &port.HostIP,
+				HostPort:      &hostPort,
+				ContainerPort: &containerPort,
+				Protocol:      &protocol,
+				Name:          &port.Name,
+			})
+		}
+
+		cgroupParent = opts.CgroupParent
+	}
+
+	podSandboxConfig.Linux = generatePodSandboxLinuxConfig(pod, cgroupParent)
+
+	if len(portMappings) > 0 {
+		podSandboxConfig.PortMappings = portMappings
+	}
+
+	if len(podResourceLimits) > 0 || len(podResourceRequests) > 0 {
+		podSandboxConfig.Resources = &runtimeApi.PodSandboxResources{}
+		var podCPULimit, podMemoryLimit, podCPURequest, podMemoryRequest int64
+		for k, v := range podResourceLimits {
+			switch k {
+			case api.ResourceCPU:
+				podCPULimit = v
+			case api.ResourceMemory:
+				podMemoryLimit = v
+			}
+		}
+		for k, v := range podResourceRequests {
+			switch k {
+			case api.ResourceCPU:
+				podCPURequest = v
+			case api.ResourceMemory:
+				podMemoryRequest = v
+			}
+		}
+
+		cpuResource := &runtimeApi.ResourceRequirements{}
+		cpulimit := float64(podCPULimit) / 1000
+		cpuResource.Limits = &cpulimit
+
+		cpuRequest := float64(podCPURequest) / 1000
+		cpuResource.Requests = &cpuRequest
+		podSandboxConfig.Resources.Cpu = cpuResource
+
+		memoryResource := &runtimeApi.ResourceRequirements{}
+		memLimit := float64(podMemoryLimit) / 1000
+		memoryResource.Limits = &memLimit
+
+		memRequest := float64(podMemoryRequest) / 1000
+		memoryResource.Requests = &memRequest
+		podSandboxConfig.Resources.Memory = memoryResource
+
+	}
+
+	return podSandboxConfig, nil
+}
+
+// generatePodSandboxLinuxConfig generates LinuxPodSandboxConfig from api.Pod.
+func generatePodSandboxLinuxConfig(pod *api.Pod, cgroupParent string) *runtimeApi.LinuxPodSandboxConfig {
+	var linuxPodSandboxConfig *runtimeApi.LinuxPodSandboxConfig
```

Comment:
I think it would be cleaner if we check if we want to return nil at the beginning.